src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -15,14 +15,70 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
         {
             var expression = BindExpression(node.Expression, diagnostics);
             var hasErrors = node.HasErrors || IsOperandErrors(node, expression, diagnostics);
-            var pattern = BindPattern(node.Pattern, expression, expression.Type, hasErrors, diagnostics);
+            var pattern = BindPattern(node.Pattern, expression, ExpressionIsNull(expression), expression.Type, hasErrors, diagnostics);
             return new BoundIsPatternExpression(
                 node, expression, pattern, GetSpecialType(SpecialType.System_Boolean, diagnostics, node), hasErrors);
         }
 
+        /// <summary>
+        /// Returns true if the expression is known to always be null.
+        /// Returns false if the expression is always known to be non-null.
+        /// Returns null otherwise.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
+        internal static bool? ExpressionIsNull(BoundExpression expression)
+        {
+            if (expression.ConstantValue != null)
+            {
+                return expression.ConstantValue.IsNull;
+            }
+
+            if (expression.Type == null || expression.Type.IsNonNullableValueType())
+            {
+                return false;
+            }
+
+            switch (expression.Kind)
```

Comment:
Yes, and that fact is recorded in its ConstantValue, and handled above.

---
In reply to: [64962722](https://github.com/dotnet/roslyn/pull/11485#discussion_r64962722) [](ancestors = 64962722)